File: app/src/main/java/com/fstyle/structure_android/data/source/remote/api/error/RequestError.java
package com.fstyle.structure_android.data.source.remote.api.error;

import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Consumer;

/**
 * Created by Sun on 4/16/2017.
 */

/**
 * Created by Sun on 4/16/2017.
 */

public abstract class RequestError implements Consumer<Throwable> {

    /**
     * Don't override this method.
     * Override {@link RequestError#onRequestError(BaseException)} instead
     */
    @Override
    public void accept(@NonNull Throwable throwable) throws Exception {
        if (throwable == null) {
            onRequestError(BaseException.toUnexpectedError(new Throwable("Unknown exception")));
            return;
        }
        if (throwable instanceof BaseException) {
            onRequestError((BaseException) throwable);
        } else {
            onRequestError(BaseException.toUnexpectedError(throwable));
        }
    }
    public abstract void onRequestError(BaseException error);
}

File: app/src/main/java/com/fstyle/structure_android/data/source/remote/api/error/RequestError.kt
package com.fstyle.structure_android.data.source.remote.api.error

import android.support.annotation.NonNull
import io.reactivex.functions.Consumer

/**
 * Created by Sun on 4/16/2017.
 */

/**
 * Created by Sun on 4/16/2017.
 */

abstract class RequestError : Consumer<Throwable> {

  /**
   * Don't override this method.
   * Override [RequestError.onRequestError] instead
   */
  @Throws(Exception::class)
  override fun accept(@NonNull throwable: Throwable) {
    if (throwable is BaseException) {
      onRequestError(throwable)
    } else {
      onRequestError(BaseException.toUnexpectedError(throwable))
    }
  }

  abstract fun onRequestError(error: BaseException)
}
